src/funcs/call-model.ts

```diff
@@ -36,32 +31,123 @@ export type CallModelInput = Omit<models.OpenResponsesRequest, 'stream' | 'tools
+ * Functions are resolved at the START of each turn in this order:
+ * 1. Async functions (parallel resolution)
+ * 2. Tool execution (if tools called by model)
+ * 3. nextTurnParams functions (if defined on tools)
+ * 4. API request with resolved values
+ *
+ * **Stop Conditions:**
+ *
+ * Control when tool execution stops using the `stopWhen` parameter:
+ *
+ * @example
+ * ```typescript
+ * // Stop after 3 steps
+ * stopWhen: stepCountIs(3)
+ *
+ * // Stop when a specific tool is called
+ * stopWhen: hasToolCall('finalizeResults')
+ *
+ * // Multiple conditions (OR logic - stops if ANY is true)
+ * stopWhen: [
+ *   stepCountIs(10),        // Safety: max 10 steps
+ *   maxCost(0.50),          // Budget: max $0.50
+ *   hasToolCall('finalize') // Logic: stop when finalize called
+ * ]
+ *
+ * // Custom condition with full step history
+ * stopWhen: ({ steps }) => {
+ *   const totalCalls = steps.reduce((sum, s) => sum + s.toolCalls.length, 0);
+ *   return totalCalls >= 20; // Stop after 20 total tool calls
+ * }
+ * ```
+ *
+ * Available helper functions:
+ * - `stepCountIs(n)` - Stop after n steps
+ * - `hasToolCall(name)` - Stop when tool is called
+ * - `maxTokensUsed(n)` - Stop when token usage exceeds n
+ * - `maxCost(n)` - Stop when cost exceeds n dollars
+ * - `finishReasonIs(reason)` - Stop on specific finish reason
+ *
+ * Default: `stepCountIs(5)` if not specified
  */
-export function callModel(
+export function callModel<TOOLS extends readonly Tool[]>(
   client: OpenRouterCore,
-  request: CallModelInput,
+  request: CallModelInput<TOOLS>,
   options?: RequestOptions,
-): ModelResult {
-  const { tools, maxToolRounds, ...apiRequest } = request;
+): ModelResult<TOOLS> {
+  const { tools, stopWhen, ...apiRequest } = request;
 
-  // Convert tools to API format and extract enhanced tools if present
+  // Convert tools to API format - no cast needed now that convertToolsToAPIFormat accepts readonly
   const apiTools = tools ? convertToolsToAPIFormat(tools) : undefined;
 
   // Build the request with converted tools
-  const finalRequest: models.OpenResponsesRequest = {
+  // Note: async functions are resolved later in ModelResult.executeToolsIfNeeded()
+  // The request can have async fields (functions) or sync fields, and the tools are converted to API format
+  const finalRequest: Record<string, unknown> = {
     ...apiRequest,
-    ...(apiTools !== undefined && {
-      tools: apiTools,
-    }),
   };
 
-  return new ModelResult({
+  if (apiTools !== undefined) {
+    finalRequest['tools'] = apiTools;
+  }
+
+  return new ModelResult<TOOLS>({
     client,
     request: finalRequest,
     options: options ?? {},
-    tools: tools ?? [],
-    ...(maxToolRounds !== undefined && {
-      maxToolRounds,
+    // Preserve the exact TOOLS type instead of widening to Tool[]
+    tools: tools as TOOLS | undefined,
+    ...(stopWhen !== undefined && {
+      stopWhen,
     }),
-  });
+  } as GetResponseOptions<TOOLS>);
```

Comment:
Why does it have to ast here btw (?)

Reply:
simpler types then interpolating between readonly and not for tools